src/host/outputStream.cpp

```diff
@@ -355,6 +356,34 @@ void ConhostInternalGetSet::SetWorkingDirectory(const std::wstring_view /*uri*/)
 {
 }
 
+// Routine Description:
+// - Plays a single MIDI note, blocking for the duration.
+// Arguments:
+// - noteNumber - The MIDI note number to be played (0 - 127).
+// - velocity - The force with which the note should be played (0 - 127).
+// - duration - How long the note should be sustained (in milliseconds).
+// Return value:
+// - true if successful. false otherwise.
+void ConhostInternalGetSet::PlayMidiNote(const int noteNumber, const int velocity, const std::chrono::microseconds duration)
+{
+    // We create the audio instance on demand, and lock it for the duration
+    // of the note output so it can't be destroyed while in use.
+    auto& midiAudio = ServiceLocator::LocateGlobals().getConsoleInformation().GetMidiAudio();
+    midiAudio.Lock();
+
+    // We then unlock the console, so the UI doesn't hang while we're busy.
+    UnlockConsole();
+
+    // This call will block for the duration, unless shutdown early.
+    midiAudio.PlayNote(noteNumber, velocity, duration);
+
+    // Once complete, we reacquire the console lock and unlock the audio.
+    // If the console has shutdown in the meantime, the Unlock call

```

Comment:
What happens if anything *else* happens while the note was playing? Are we resizing, reflowing, allowing the user to lock the console with a pending selection or open the property sheet? If so, how do each of those things impact what happens next?

Reply:
That's a good question. In conhost it should be fine I think. When we release the lock, it's similar to what would happen if we'd just reached the end of the output buffer at that point. And then when we reacquire the lock, that's similar to receiving a new chunk of content that needs to be processed.

It gets trickier with a conpty connection though. Conpty doesn't know that it's blocked on the other side, so when you resize, I think it still sends through frame updates to redraw the page. But because those are blocked, they build up and produce a bit of a mess when they finally get through.

You can test with a sequence like this, which play 16 seconds worth of notes in a single call:

	printf "\e[4;32;1;2;3;4;5;6;7;8;9;10;11;12;13;14;15;16,~"

I suspect that part of the problem is bash wanting to redraw the prompt when we resize, which we don't always handle very well at the best of times. But it just looks a little messed up - it's not like it's going to kill anything (I hope).

Also worth mentioning that the passthrough mode doesn't have this problem as far as I can see.